compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -263,56 +258,103 @@ impl<'hir> LoweringContext<'_, 'hir> {
+            let final_expr = this.finalize_body_lowering(delegation, target_expr, args, span);
             (this.arena.alloc_from_iter(parameters), final_expr)
         })
     }
 
-    fn generate_call(
+    // Generates expression for the resulting body. If possible, `MethodCall` is used
+    // instead of fully qualified call for the self type coercion. See `consider_candidates`
+    // for more information.
+    fn finalize_body_lowering(
         &mut self,
-        path: hir::QPath<'hir>,
-        args: &'hir [hir::Expr<'hir>],
+        delegation: &Delegation,
+        target_expr: Option<hir::Block<'hir>>,
+        mut args: Vec<hir::Expr<'hir>>,
+        span: Span,
     ) -> hir::Expr<'hir> {
-        let callee = self.arena.alloc(hir::Expr {
-            hir_id: self.next_id(),
-            kind: hir::ExprKind::Path(path),
-            span: path.span(),
-        });
+        let (stmts, call, block_id) = if self
+            .get_resolution_id(delegation.id, span)
+            .and_then(|def_id| Ok(self.has_self(def_id, span)))
+            .unwrap_or_default()
+            && delegation.qself.is_none()
+        {
+            let ast_segment = delegation.path.segments.last().unwrap();
+            let segment = self.lower_path_segment(
+                delegation.path.span,
+                ast_segment,
+                ParamMode::Optional,
+                ParenthesizedGenericArgs::Err,
+                ImplTraitContext::Disallowed(ImplTraitPosition::Path),
+                None,
+                None,
+            );
+            let segment = self.arena.alloc(segment);
+
+            let args = &*self.arena.alloc_from_iter(args);
+            let (stmts, receiver, args, block_id) = if let Some(target_expr) = target_expr {
+                // Use unit type as a receiver if no expression is provided.
+                let receiver = target_expr.expr.unwrap_or_else(|| {
+                    self.arena.alloc(self.mk_expr(hir::ExprKind::Tup(&[]), span))
+                });
+                (target_expr.stmts, receiver, args, target_expr.hir_id)
+            } else {
+                (&*self.arena.alloc_from_iter([]), &args[0], &args[1..], self.next_id())
+            };
 
-        let expr = self.arena.alloc(hir::Expr {
-            hir_id: self.next_id(),
-            kind: hir::ExprKind::Call(callee, args),
-            span: path.span(),
-        });
+            let method_call_id = self.next_id();
+            if let Some(traits) = self.resolver.delegation_trait_map.remove(&delegation.id) {
+                self.trait_map.insert(method_call_id.local_id, traits.into_boxed_slice());
+            }
+
+            let method_call = self.arena.alloc(hir::Expr {
+                hir_id: method_call_id,
+                kind: hir::ExprKind::MethodCall(segment, receiver, args, span),
+                span,
+            });
+
+            (stmts, method_call, block_id)
+        } else {
+            let path = self.lower_qpath(
```

Comment:
It's hard to read the delegation.rs changes because refactoring changes like `mk_expr`, the move of `lower_qpath`, or extraction of `has_self` are merged together with the functional changes.

Reply:
`mk_expr` is the only refactoring change. `lower_qpath` should be called only when lowering to fully qualified call. Otherwise `index_hir` will emit an error. `has_self` is necessary to check both `res_id` and `sig_id`.

Reply:
refactoring changes were moved to another PR: https://github.com/rust-lang/rust/pull/126947

Reply:
I've r+'d https://github.com/rust-lang/rust/pull/126947, but a separate commit in this PR would be enough.